cterasdk/core/templates.py

```diff
@@ -60,12 +67,52 @@ def add(self, name, description=None, include_sets=None, exclude_sets=None):
                 param.deviceSettings.backup.backupPolicy.includeSets = include_sets
             if exclude_sets:
                 param.deviceSettings.backup.backupPolicy.excludeSets = exclude_sets
+            if backup_schedule:
+                param.deviceSettings.backup.scheduleTopic = Object()
+                param.deviceSettings.backup.scheduleTopic._classname = 'ScheduleTopic'  # pylint: disable=protected-access
+                param.deviceSettings.backup.scheduleTopic.overrideTemplate = True
+                param.deviceSettings.backup.scheduleTopic.schedule = backup_schedule
+            if apps:
+                param.deviceSettings.backup.applicationsTopic = Object()
+                param.deviceSettings.backup.applicationsTopic._classname = 'ApplicationsTopic'  # pylint: disable=protected-access
+                param.deviceSettings.backup.applicationsTopic.overrideTemplate = True
+                param.deviceSettings.backup.applicationsTopic.includeApps = ApplicationBackupSet(apps)
+
+        param.firmwaresSettings = Object()
+        param.firmwaresSettings._classname = 'FirmwaresSettings'  # pylint: disable=protected-access
+
+        if versions:
+            param.firmwaresSettings.useGlobal = False
+            param.firmwaresSettings.firmwares = self._convert_to_template_firmwares(versions)
```

Comment:
Since this method can fail, move the call to the start (so you fail faster). In addition, consider moving the creation of the firmwareSettings object to a separate method